src/sys/sendfile.rs

```diff
@@ -6,8 +6,98 @@ use libc::{self, off_t};
 use Result;
 use errno::Errno;
 
+#[cfg(any(target_os = "linux",
+          target_os = "android",
+          all(unix, target_env = "uclibc")))]
 pub fn sendfile(out_fd: RawFd, in_fd: RawFd, offset: Option<&mut off_t>, count: usize) -> Result<usize> {
     let offset = offset.map(|offset| offset as *mut _).unwrap_or(ptr::null_mut());
     let ret = unsafe { libc::sendfile(out_fd, in_fd, offset, count) };
     Errno::result(ret).map(|r| r as usize)
 }
+
+cfg_if! {
+    if #[cfg(any(target_os = "dragonfly",
+                 target_os = "freebsd",
+                 target_os = "ios",
+                 target_os = "macos"))] {
+        use std::marker::PhantomData;
+        use sys::uio::IoVec;
+
+        #[allow(missing_debug_implementations)]
+        #[repr(C)]
+        pub struct SendfileHeaderTrailer<'a>(libc::sf_hdtr, PhantomData<&'a [IoVec<&'a [u8]>]>);
+
+        impl<'a> SendfileHeaderTrailer<'a> {
+            pub fn new(headers: &'a [IoVec<&'a [u8]>], trailers: &'a [IoVec<&'a [u8]>]) -> SendfileHeaderTrailer<'a> {
+                SendfileHeaderTrailer(libc::sf_hdtr {
+                    headers: { if !headers.is_empty() { headers.as_ptr() as *mut libc::iovec } else { ptr::null_mut() } },
+                    hdr_cnt: headers.len() as i32,
+                    trailers: { if !trailers.is_empty() { trailers.as_ptr() as *mut libc::iovec} else { ptr::null_mut() } },
+                    trl_cnt: trailers.len() as i32
+                }, PhantomData)
+            }
+        }
+    }
+}
+
+cfg_if! {
+    if #[cfg(target_os = "freebsd")] {
+        use libc::c_int;
+
+        libc_bitflags!{
+            pub struct SfFlags: c_int {
+                SF_NODISKIO;
+                SF_SYNC;
+                // SF_USER_READAHEAD; // TODO: waiting on libc release
```

Comment:
Nix uses a git dependency on libc (except when we release), so you can go ahead and enable these now.